bindings/go/dag/sync/discover.go

```diff
@@ -31,164 +31,155 @@ func (s DiscoveryState) String() string {
-	// DiscoveryStateDiscovered indicates the vertex has been processed by the
-	// DiscoverNeighbors, but its neighbors or transitive neighbors have not all
-	// been processed by DiscoverNeighbors yet.
+// Discovery states describe how far processing has gone for each vertex.
+const (
+	// DiscoveryStateUnknown means the vertex has not been processed yet or is in an unknown state.
+	DiscoveryStateUnknown DiscoveryState = iota
+	// DiscoveryStateDiscovering means the vertex is being resolved.
+	DiscoveryStateDiscovering
+	// DiscoveryStateDiscovered means the vertex was resolved but neighbors were not yet fully explored.
 	DiscoveryStateDiscovered
-	// DiscoveryStateCompleted indicates the vertex and all its neighbors have been
-	// processed by the DiscoverNeighbors (sub-graph up to this vertex is fully
-	// completed).
+	// DiscoveryStateCompleted means the vertex and all its reachable neighbors successfully discovered.
 	DiscoveryStateCompleted
-	// DiscoveryStateError indicates DiscoverNeighbors returned an error for this vertex
-	// or a neighbor.
+	// DiscoveryStateError means discovery failed for this vertex or one of its neighbors.
 	DiscoveryStateError
 )
 
-// TODO(fabianburth): Add a recursion depth limit
-type DiscoverOptions[T cmp.Ordered] struct {
-	// Roots to start the discovery from
-	Roots          []T
-	GoRoutineLimit int
+// NewGraphDiscoverer constructs a discoverer instance with its own
+// concurrency-safe DAG and per-vertex "done" channels.
+func NewGraphDiscoverer[K cmp.Ordered, V any](opts *GraphDiscovererOptions[K, V]) *GraphDiscoverer[K, V] {
+	return &GraphDiscoverer[K, V]{
+		graph:   NewSyncedDirectedAcyclicGraph[K](),
+		doneMap: &sync.Map{},
+		opts:    opts,
+	}
 }
 
-type DiscoverOption[T cmp.Ordered] func(*DiscoverOptions[T])
+// GraphDiscoverer orchestrates concurrent graph discovery.
+// - graph: the shared DAG being built
+// - doneMap: ensures each vertex is processed exactly once
+// - opts: external resolver and neighbor discoverer
+type GraphDiscoverer[K cmp.Ordered, V any] struct {
+	opts    *GraphDiscovererOptions[K, V]
+	graph   *SyncedDirectedAcyclicGraph[K]
+	doneMap *sync.Map
+}
 
-func WithDiscoveryGoRoutineLimit[T cmp.Ordered](numGoRoutines int) DiscoverOption[T] {
-	return func(options *DiscoverOptions[T]) {
-		options.GoRoutineLimit = numGoRoutines
-	}
+func (d *GraphDiscoverer[K, V]) Graph() *SyncedDirectedAcyclicGraph[K] {
+	return d.graph
 }
 
-func WithRoots[T cmp.Ordered](root ...T) DiscoverOption[T] {
-	return func(options *DiscoverOptions[T]) {
-		options.Roots = root
-	}
+// CurrentEdges returns the ordered neighbors of a vertex from the DAG snapshot.
+func (d *GraphDiscoverer[K, V]) CurrentEdges(key K) []K {
+	var edges []K
+	_ = d.graph.WithReadLock(func(d *dag.DirectedAcyclicGraph[K]) error {
+		v, ok := d.Vertices[key]
+		if !ok {
+			return nil
+		}
+		edges = make([]K, len(v.Edges))
+		for k, edge := range v.Edges {
+			idx := edge[attributeOrderIndex]
```

Comment:
Do an exists check here? Otherwise, it could end up being a problem in the below `idx.(int)`.

Reply:
how should I react if we do not have the index?